### ISSUE TITLE
PF4: Virtualized Table/Virtualized Extension

### ISSUE THREAD

**Issue by priley86** [patternfly/patternfly-react#1229]
This issue builds on to #402 for adding the virtualized feature to PF4 React tables.

Relevant design/core issues:
None as of now (may not be needed)

We would like to release this feature as a "pf4 extensions" package right now. This functionality could be distributed and labeled as `@patternfly/react-virtualized-extension` for the time being so that product teams can consume now and revisit later when design review is completed. The utilities developed in this package could be useful for tables, lists, and other scrolling components.

Example from Reactabular:
https://reactabular.js.org/#/features/virtualization

**Comment by ohadlevy** [comment]
may I ask what was the reason for infinite scrolling? AFAIR we moved away
from it in other projects due to bad usability with filtering/searching and
performance?

**Comment by priley86** [comment]
@ohadlevy - thanks for speaking up - it's a good time to consider this 😸. Maybe you can elaborate? 

I believe PF design direction and direction from OpenShift yesterday was to continue supporting infinite scroll/virtualized list. The intention is to improve performance by reusing existing dom elements as the user scrolls the long list. To accomplish this w/ Reactabular, utility methods could be extracted from the [react-virtualized](https://github.com/bvaughn/react-virtualized) project currently in use or the [reactabular-virtualized](https://github.com/reactabular/reactabular/tree/master/packages/reactabular-virtualized) project provided  by Reactabular (still researching this).

cc: @janwright73 @serenamarie125 @lizsurette - if we decide infinite scroll is no longer favorable, we could look to other design patterns, i.e. pagination, "load more" buttons, etc. Infinite scroll was the currently requested option though.

**Comment by priley86** [comment]
I think with this issue it is important to distinguish infinite scroll separately (could be tackled in another issue). This issue would be focused on providing utility methods for reusing the dom and calculating row heights (virtualizing). A separate issue for providing a loading indicator/asynchronous paging of contents could be implemented alongside or as a separate pattern. 

https://uxplanet.org/ux-infinite-scrolling-vs-pagination-1030d29376f1

**Comment by serenamarie125** [comment]
> may I ask what was the reason for infinite scrolling? AFAIR we moved away from it in other projects due to bad usability with filtering/searching and performance?

@ohadlevy curious what the usability issues were that you encountered ( other than performance )?

I think that PF suggests using pagination if there are constraints in the back end which present performance issues.  

Historically we have seen pagination present a number of usability issues, including issues with selection across multiple pages, filter and/or sort not properly working across the entire data set, and find/filter quickly by name not being able to work across pages.  These problems are not an issue with this pattern.

**Comment by priley86** [comment]
I've posted a few links in the "patternfly-a11y" channel pertaining to this topic. Part of me thinks we can support all of the same use cases w/ client side pagination in OpenShift (sort of backtracking on this now). You would sill be able to deep link pages/filters/sorts into the URL with a client paginated table, and sort or filter across the entire data set (assuming it remains client paginated and not server paginated). 

The ability to deep link position in the infinitely scrolled table is not there today, but could be added. It becomes quite trivial w/ a paginated table though. 

The deeper concern after iterating in #1250 I have is more w/ keyboard accessibility of the pages in the table.

More on that in these articles I've referenced:
https://www.digitala11y.com/infinite-scroll-accessibility-is-it-any-good/
https://www.w3.org/TR/wai-aria-practices-1.1/examples/feed/feed.html
